sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/ThroughputControlStore.java

```diff
@@ -0,0 +1,230 @@
+                        return Mono.just(containerController);
+                    }
+
+                    // Unable to find container controller to handle the request,
+                    // It is caused by control store out of sync or the request has staled info.
+                    // We will handle the first scenario by creating a new container controller,
+                    // while fall back to original request Mono for the second scenario.
+                    return this.shouldRefreshContainerController(collectionLink, request)
+                        .flatMap(shouldRefresh -> {
+                            if (shouldRefresh) {
+                                containerController.close();
+                                this.containerControllerCache.refresh(collectionLink, () -> this.createAndInitContainerController(collectionLink));
+                                return this.resolveContainerController(collectionLink);
+                            }
+
+                            // The container container controller is up to date, the request has staled info, will let the request pass
+                            return Mono.just(containerController);
+                        });
+                })
+                .flatMap(containerController -> {
+                    if (containerController.canHandleRequest(request)) {
+                        return containerController.processRequest(request, originalRequestMono)
+                            .doOnError(throwable -> this.doOnError(request, containerController, throwable));
+                    } else {
+                        // still can not handle the request
+                        logger.warn(
+                            "Can not find container controller to process request {} with collectionRid {} ",
+                            request.getActivityId(),
+                            request.requestContext.resolvedCollectionRid);
+
+                        return originalRequestMono;
+                    }
+                });
+        } else {
+            return originalRequestMono;
+        }
+    }
+
+    private Mono<Boolean> shouldRefreshContainerController(String containerLink, RxDocumentServiceRequest request) {
+        // TODO: populate diagnostics
+        return this.collectionCache.resolveByNameAsync(null, containerLink, null)
+            .flatMap(documentCollection ->
+                Mono.just(StringUtils.equals(documentCollection.getResourceId(), request.requestContext.resolvedCollectionRid)));
+    }
+
+    private void doOnError(RxDocumentServiceRequest request, IThroughputController controller, Throwable throwable) {
+        checkNotNull(request, "Request can not be null");
+        checkNotNull(controller, "Container controller can not be null");
+        checkNotNull(throwable, "Exception can not be null");
+
+        CosmosException cosmosException = Utils.as(throwable, CosmosException.class);
```

Comment:
Better to unwrap this exception - in case the exception is wrapped in `ReactiveException` -> 

```
final Throwable throwable = Exceptions.unwrap(ex);
CosmosException cosmosException = Utils.as(throwable, CosmosException.class);
```

Reply:
Updated